File: src/desktop/modern/core/dependency_injection/lifecycle_manager.py
"""
Lifecycle Manager - Service Lifecycle and Cleanup Management

Handles:
- Service initialization
- Service cleanup and disposal
- Scoped instance management
- Lifecycle event coordination
"""

from __future__ import annotations

from collections.abc import Callable
import logging
from typing import Any


logger = logging.getLogger(__name__)


class LifecycleManager:
    """
    Manages service lifecycle including initialization, cleanup, and scoped instances.

    Provides centralized lifecycle management for all services in the DI container,
    ensuring proper initialization and cleanup of resources.
    """

    def __init__(self):
        self.cleanup_handlers: list[Callable] = []
        self._scoped_instances: dict[str, dict[type, Any]] = {}
        self._current_scope: str | None = None
        self._initialized_services: list[Any] = []

    def create_with_lifecycle(self, instance: Any) -> Any:
        """Create instance with proper lifecycle management."""
        # Call initialization method if it exists
        if hasattr(instance, "initialize") and callable(instance.initialize):
            try:
                instance.initialize()
                self._initialized_services.append(instance)
                logger.debug(f"Initialized service: {type(instance).__name__}")
            except Exception as e:
                logger.error(
                    f"Failed to initialize service {type(instance).__name__}: {e}"
                )
                raise

        # Register for cleanup if it has cleanup method
        if hasattr(instance, "cleanup") and callable(instance.cleanup):
            self.cleanup_handlers.append(instance.cleanup)
            logger.debug(f"Registered cleanup for service: {type(instance).__name__}")

        return instance

    def cleanup_all(self) -> None:
        """Cleanup all registered services."""
        logger.info(f"Starting cleanup of {len(self.cleanup_handlers)} services")

        for cleanup_handler in reversed(self.cleanup_handlers):
            try:
                cleanup_handler()
                logger.debug("Service cleanup completed successfully")
            except Exception as e:
                logger.error(f"Error during service cleanup: {e}")

        self.cleanup_handlers.clear()
        self._initialized_services.clear()
        logger.info("All services cleaned up")

    def create_scope(self, scope_id: str) -> None:
        """Create a new scope for scoped services."""
        if scope_id in self._scoped_instances:
            logger.warning(f"Scope {scope_id} already exists, overwriting")

        self._scoped_instances[scope_id] = {}
        self._current_scope = scope_id
        logger.debug(f"Created scope: {scope_id}")

    def dispose_scope(self, scope_id: str) -> None:
        """Dispose a scope and cleanup its instances."""
        if scope_id not in self._scoped_instances:
            logger.warning(f"Scope {scope_id} does not exist")
            return

        scope_instances = self._scoped_instances[scope_id]
        logger.debug(
            f"Disposing scope {scope_id} with {len(scope_instances)} instances"
        )

        # Cleanup scoped instances
        for instance in scope_instances.values():
            if hasattr(instance, "cleanup") and callable(instance.cleanup):
                try:
                    instance.cleanup()
                    logger.debug(
                        f"Cleaned up scoped instance: {type(instance).__name__}"
                    )
                except Exception as e:
                    logger.error(f"Error cleaning up scoped instance: {e}")

        del self._scoped_instances[scope_id]
        if self._current_scope == scope_id:
            self._current_scope = None

        logger.debug(f"Disposed scope: {scope_id}")

    def get_scoped_instance(self, scope_id: str, service_type: type) -> Any | None:
        """Get an instance from a specific scope."""
        if scope_id not in self._scoped_instances:
            return None
        return self._scoped_instances[scope_id].get(service_type)

    def set_scoped_instance(
        self, scope_id: str, service_type: type, instance: Any
    ) -> None:
        """Store an instance in a specific scope."""
        if scope_id not in self._scoped_instances:
            self.create_scope(scope_id)

        self._scoped_instances[scope_id][service_type] = instance
        logger.debug(
            f"Stored scoped instance: {type(instance).__name__} in scope {scope_id}"
        )

    def get_current_scope(self) -> str | None:
        """Get the current active scope."""
        return self._current_scope

    def set_current_scope(self, scope_id: str | None) -> None:
        """Set the current active scope."""
        self._current_scope = scope_id
        logger.debug(f"Set current scope to: {scope_id}")

    def has_scope(self, scope_id: str) -> bool:
        """Check if a scope exists."""
        return scope_id in self._scoped_instances

    def get_scope_instance_count(self, scope_id: str) -> int:
        """Get the number of instances in a scope."""
        if scope_id not in self._scoped_instances:
            return 0
        return len(self._scoped_instances[scope_id])

    def get_total_scoped_instances(self) -> int:
        """Get total number of scoped instances across all scopes."""
        return sum(len(instances) for instances in self._scoped_instances.values())

    def get_active_scopes(self) -> list[str]:
        """Get list of all active scope IDs."""
        return list(self._scoped_instances.keys())

    def dispose_all_scopes(self) -> None:
        """Dispose all scopes and their instances."""
        scope_ids = list(self._scoped_instances.keys())
        for scope_id in scope_ids:
            self.dispose_scope(scope_id)

        self._current_scope = None
        logger.info("All scopes disposed")

    def get_cleanup_handler_count(self) -> int:
        """Get the number of registered cleanup handlers."""
        return len(self.cleanup_handlers)

    def get_initialized_service_count(self) -> int:
        """Get the number of initialized services."""
        return len(self._initialized_services)

    def force_cleanup_service(self, service_instance: Any) -> bool:
        """Force cleanup of a specific service instance."""
        if hasattr(service_instance, "cleanup") and callable(service_instance.cleanup):
            try:
                service_instance.cleanup()

                # Remove from cleanup handlers if present
                if service_instance.cleanup in self.cleanup_handlers:
                    self.cleanup_handlers.remove(service_instance.cleanup)

                # Remove from initialized services if present
                if service_instance in self._initialized_services:
                    self._initialized_services.remove(service_instance)

                logger.debug(
                    f"Force cleaned up service: {type(service_instance).__name__}"
                )
                return True
            except Exception as e:
                logger.error(
                    f"Error during force cleanup of {type(service_instance).__name__}: {e}"
                )
                return False

        logger.warning(
            f"Service {type(service_instance).__name__} has no cleanup method"
        )
        return False

    def get_lifecycle_stats(self) -> dict[str, Any]:
        """Get comprehensive lifecycle statistics."""
        return {
            "cleanup_handlers": len(self.cleanup_handlers),
            "initialized_services": len(self._initialized_services),
            "active_scopes": len(self._scoped_instances),
            "total_scoped_instances": self.get_total_scoped_instances(),
            "current_scope": self._current_scope,
            "scope_details": {
                scope_id: len(instances)
                for scope_id, instances in self._scoped_instances.items()
            },
        }

File: src/desktop/shared/di/core/LifecycleManager.ts
/**
 * 🔄 TKA LIFECYCLE MANAGER
 *
 * Advanced service lifecycle management with sophisticated scoping,
 * disposal patterns, and resource cleanup.
 */

import { ServiceScope, ResolutionContext, IDisposable, IInitializable } from './types.js';

export class LifecycleManager {
    private readonly _scopes = new Map<string, ScopeManager>();
    private readonly _disposables = new Set<IDisposable>();
    private readonly _initializables = new Set<IInitializable>();
    private _currentScope?: string;

    /**
     * Create a new scope for scoped services
     */
    createScope(scopeId: string): void {
        if (this._scopes.has(scopeId)) {
            throw new Error(`Scope '${scopeId}' already exists`);
        }

        const scopeManager = new ScopeManager(scopeId);
        this._scopes.set(scopeId, scopeManager);
    }

    /**
     * Dispose a scope and cleanup its instances
     */
    disposeScope(scopeId: string): void {
        const scopeManager = this._scopes.get(scopeId);
        if (scopeManager) {
            scopeManager.dispose();
            this._scopes.delete(scopeId);
        }

        if (this._currentScope === scopeId) {
            this._currentScope = undefined;
        }
    }

    /**
     * Set the current scope for scoped service resolution
     */
    setCurrentScope(scopeId: string): void {
        if (!this._scopes.has(scopeId)) {
            throw new Error(`Scope '${scopeId}' does not exist`);
        }
        this._currentScope = scopeId;
    }

    /**
     * Get the current scope ID
     */
    getCurrentScope(): string | undefined {
        return this._currentScope;
    }

    /**
     * Apply lifecycle management to a service instance
     */
    createWithLifecycle<T>(instance: T, context: ResolutionContext): T {
        // Initialize if needed
        if (this._isInitializable(instance)) {
            this._initializables.add(instance);
            this._initializeInstance(instance);
        }

        // Track disposables
        if (this._isDisposable(instance)) {
            this._disposables.add(instance);
        }

        // Apply scope-specific lifecycle
        this._applyScopeLifecycle(instance, context);

        return instance;
    }

    /**
     * Get lifecycle information for debugging
     */
    getLifecycleInfo() {
        return {
            scopes: Array.from(this._scopes.keys()),
            currentScope: this._currentScope,
            disposables: this._disposables.size,
            initializables: this._initializables.size,
            scopeDetails: Object.fromEntries(
                Array.from(this._scopes.entries()).map(([id, manager]) => [
                    id,
                    manager.getInfo()
                ])
            )
        };
    }

    /**
     * Dispose all managed resources
     */
    dispose(): void {
        // Dispose all scopes
        for (const scopeManager of this._scopes.values()) {
            scopeManager.dispose();
        }
        this._scopes.clear();

        // Dispose all tracked disposables
        for (const disposable of this._disposables) {
            try {
                disposable.dispose();
            } catch (error) {
                console.error('Error disposing service:', error);
            }
        }
        this._disposables.clear();
        this._initializables.clear();
        this._currentScope = undefined;
    }

    private _isDisposable(instance: any): instance is IDisposable {
        return instance && typeof instance.dispose === 'function';
    }

    private _isInitializable(instance: any): instance is IInitializable {
        return instance && typeof instance.initialize === 'function';
    }

    private async _initializeInstance(instance: IInitializable): Promise<void> {
        try {
            const result = instance.initialize();
            if (result instanceof Promise) {
                await result;
            }
        } catch (error) {
            console.error('Error initializing service:', error);
        }
    }

    private _applyScopeLifecycle<T>(instance: T, context: ResolutionContext): void {
        const scopeId = this._determineScopeId(context);
        if (scopeId) {
            const scopeManager = this._scopes.get(scopeId);
            if (scopeManager) {
                scopeManager.addInstance(instance);
            }
        }
    }

    private _determineScopeId(context: ResolutionContext): string | undefined {
        if (context.scopeId) return context.scopeId;
        if (context.requestId) return `request_${context.requestId}`;
        if (context.sessionId) return `session_${context.sessionId}`;
        if (context.componentId) return `component_${context.componentId}`;
        return this._currentScope;
    }
}

/**
 * Manages instances within a specific scope
 */
class ScopeManager {
    private readonly _scopeId: string;
    private readonly _instances = new Set<any>();
    private readonly _createdAt = new Date();
    private _disposed = false;

    constructor(scopeId: string) {
        this._scopeId = scopeId;
    }

    addInstance(instance: any): void {
        if (this._disposed) {
            throw new Error(`Scope '${this._scopeId}' has been disposed`);
        }
        this._instances.add(instance);
    }

    removeInstance(instance: any): void {
        this._instances.delete(instance);
    }

    getInfo() {
        return {
            scopeId: this._scopeId,
            instanceCount: this._instances.size,
            createdAt: this._createdAt,
            disposed: this._disposed
        };
    }

    dispose(): void {
        if (this._disposed) return;

        // Dispose all instances that are disposable
        for (const instance of this._instances) {
            if (instance && typeof instance.dispose === 'function') {
                try {
                    instance.dispose();
                } catch (error) {
                    console.error(`Error disposing instance in scope '${this._scopeId}':`, error);
                }
            }
        }

        this._instances.clear();
        this._disposed = true;
    }
}
